File: sample-videochat-java/app/src/main/java/com/quickblox/sample/videochat/java/executor/Executor.java
package com.quickblox.sample.videochat.java.executor;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.BlockingQueue;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

public class Executor {
    private final int CORE_POOL_SIZE = 3;
    private final int MAX_POOL_SIZE = 3;
    private final int KEEP_ALIVE_TIME = 1;

    private final BlockingQueue<Runnable> threadQueue = new LinkedBlockingQueue<>();
    private final ThreadPoolExecutor threadPoolExecutor = new ThreadPoolExecutor(CORE_POOL_SIZE, MAX_POOL_SIZE,
            KEEP_ALIVE_TIME, TimeUnit.SECONDS, threadQueue);

    private static Executor instance;

    private static synchronized Executor getInstance() {
        if (instance == null) {
            instance = new Executor();
        }
        return instance;
    }

    private Executor() {
    }

    public static <T> void addTask(ExecutorTask<T> executorTask) {
        getInstance().threadPoolExecutor.execute(() -> {
            Handler mainHandler = new Handler(Looper.getMainLooper());

            try {
                T result = executorTask.onBackground();
                mainHandler.post(() -> executorTask.onForeground(result));
            } catch (Exception exception) {
                mainHandler.post(() -> executorTask.onError(exception));
            }
        });
    }
}

File: sample-videochat-kotlin/app/src/main/java/com/quickblox/sample/videochat/kotlin/executor/Executor.kt
package com.quickblox.sample.videochat.kotlin.executor

import android.os.Handler
import android.os.Looper
import java.util.concurrent.BlockingQueue
import java.util.concurrent.LinkedBlockingQueue
import java.util.concurrent.ThreadPoolExecutor
import java.util.concurrent.TimeUnit

private const val THREAD_POOL_SIZE = 3
private const val MAX_POOL_SIZE = 3
private const val KEEP_ALIVE_TIME = 1L

object Executor {
    private var threadPoolExecutor: ThreadPoolExecutor? = null
    private val threadQueue: BlockingQueue<Runnable> = LinkedBlockingQueue()

    init {
        threadPoolExecutor = ThreadPoolExecutor(
            THREAD_POOL_SIZE,
            MAX_POOL_SIZE,
            KEEP_ALIVE_TIME,
            TimeUnit.SECONDS,
            threadQueue
        )
    }

    fun <T> addTask(executorTask: ExecutorTask<T>) {
        threadPoolExecutor?.execute {
            val mainHandler = Handler(Looper.getMainLooper())

            try {
                val result = executorTask.onBackground()
                mainHandler.post {
                    executorTask.onForeground(result)
                }
            } catch (exception: Exception) {
                mainHandler.post {
                    executorTask.onError(exception)
                }
            }
        }
    }
}